proposals/0055-transact-without-require-weight-at-most.md

```diff
@@ -0,0 +1,65 @@
+---
+Title: Transact without `require_weight_at_most`
+Number: 55
+Status: Draft
+Version: 0
+Authors:
+ - Adrian Catangiu
+ - Francisco Aguirre
+Created: 2024-04-25
+Impact: High
+Requires:
+Replaces:
+---
+
+## Summary
+
+The Transact instruction shouldn't need the `require_weight_at_most` parameter, since the decoder could just get the weight of the call from its dispatch info.
+
+## Motivation
+
+The UX of using `Transact` is not great, and one part of the problem is guesstimating this `require_weight_at_most`.
+We've seen multiple `Transact`s on-chain failures caused by the "incorrect" use or the parameter. In practice,
+this parameter only adds UX overhead. The vast majority of use cases fall in one of two categories:
+1. Unpaid execution of Transacts - in these cases the `require_weight_at_most` is not really useful, caller doesn't 
+have to pay for it, and on the call site it either fits the block or not;
+2. Paid execution of _single_ Transact - the weight to be spent by the Transact is already covered by the `BuyExecution`
+weight limit parameter.
+
+We've had multiple OpenGov `root/whitelisted_caller` proposals initiated by core-devs completely or partially fail
+because of incorrect configuration of `require_weight_at_most` parameter. This is a strong indication that the
+instruction is hard to use.
+
+## Specification
+
+The proposal is simple: change `Transact` instruction:
+
+```diff
+- Transact { origin_kind: OriginKind, require_weight_at_most: Weight, call: DoubleEncoded<Call> },
++ Transact { origin_kind: OriginKind, maybe_weight_limit: Option<WeightLimit>, call: DoubleEncoded<Call> },
+```
+
+With this new API, users who do not need to artificially limit the maximum weight used by the inner `call`,
+can pass `maybe_weight_limit: None`; while those who need to do it, still can.
+
+## Security considerations
+
+TODO
```

Comment:
Currently, an XCVM implementation can weigh a message just by looking at the decoded instructions without decoding the Transact's `call`. Is it safe to decode the `call` during the message weighing (i.e., during a "free" execution) to get the dispatch info?

Reply:
Conceptually not that much changes. Time for decoding is not included in the weight right now or maybe part of the base weight or similar. 

However, right now you could already say that execution takes `5Weight` and then it continues. When you come to the point of applying the call you decode and realize that it actually requires `6Weight` and you bail. So, you also decode it in the current model for "free". Discussable if it was for free, because you already paid fees, but they are paid based on the weight.

Reply:
Considering [the decoding is limited by `MAX_XCM_DECODE_DEPTH`](https://github.com/paritytech/polkadot-sdk/blob/release-crates-io-v1.7.0/polkadot/xcm/src/double_encoded.rs#L94), even if the decoding was "free" (`Transact` benchmarking is using the smallest possible call), it does have a limit on how much computation it can take. If you want to be extremely cautious about it, you could benchmark the `Transact` instruction using a call close to `MAX_XCM_DECODE_DEPTH` (max weight usage), and this will reduce the "free" side of it. However, as the difference in weight are not significant currently, you can consider it is safe to decode the call using weights generated with the current beckmarking.

Reply:
I've updated the "security considerations" section.

The `MAX_XCM_DECODE_DEPTH` is IMO an implementation detail and should not be part of the formal spec, so I haven't included it.